# Exploratory data analysis of telco customer data

In [1]:
# Importing modules
import pandas as pd

In [2]:
# Loading data to dataframe
df = pd.read_csv('Telco Customer Churn.csv')

In [3]:
# Checking for null values
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

- No missing / null value found

In [4]:
# Gathering info about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


- The dataset contains data of 7043 customers.
- TotalCharges column contains numeric values but its data type is object. We shall convert it to numeric.

In [5]:
# Converting data type of 'TotalCharges' column to float
df['TotalCharges'] = df['TotalCharges'].astype(float)

ValueError: could not convert string to float: ''

- Could not convert data type to float. Empty strings (' ') are present which can't be converted to float. 

In [6]:
# Counting the no. of empty string values
count = 0
for i in df['TotalCharges']:
    if i ==' ':
        count+=1
print(count)

11


- There are 11 out of 7043 rows where a value is missing. As it is a small number we can remove these rows as there will not be any major impact on the analysis.

In [7]:
# Dropping rows containing empty strings
df = df.replace(' ', pd.NA).dropna()

In [8]:
# Converting data type of 'TotalCharges' column to float
df['TotalCharges'] = df['TotalCharges'].astype(float)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 
 17  

- TotalCharges has been converted to float type.

In [10]:
# Checking for duplicate data
print(len(pd.unique(df['customerID'])))

7032


- All customers are unique. No duplicate customerID found.

In [11]:
# Description of data
df.describe()

SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7032.000000  7032.000000     7032.000000   7032.000000
mean        0.162400    32.421786       64.798208   2283.300441
std         0.368844    24.545260       30.085974   2266.771362
min         0.000000     1.000000       18.250000     18.800000
25%         0.000000     9.000000       35.587500    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.862500   3794.737500
max         1.000000    72.000000      118.750000   8684.800000

- 16.2% of all customers are senior citizens.
- Average tenure of a customer is 32.4 months and max tenure is 72 months.
- Average monthly charge paid by customers is ```$64.76```. Min paid charge is ```$18.25``` while max is ```$118.75```.

In [12]:
# Displaying first 10 rows of dataframe to know more about the dateset
df.head(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   
5  9305-CDSKC  Female              0      No         No       8          Yes   
6  1452-KIOVK    Male              0      No        Yes      22          Yes   
7  6713-OKOMC  Female              0      No         No      10           No   
8  7892-POOKP  Female              0     Yes         No      28          Yes   
9  6388-TABGU    Male              0      No        Yes      62          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   
5               Yes     Fiber optic             No  ...              Yes   
6               Yes     Fiber optic             No  ...               No   
7  No phone service             DSL            Yes  ...               No   
8               Yes     Fiber optic             No  ...              Yes   
9                No             DSL            Yes  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   
5          No         Yes             Yes  Month-to-month              Yes   
6          No         Yes              No  Month-to-month              Yes   
7          No          No              No  Month-to-month               No   
8         Yes         Yes             Yes  Month-to-month              Yes   
9          No          No              No        One year               No   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  
5           Electronic check          99.65        820.50    Yes  
6    Credit card (automatic)          89.10       1949.40     No  
7               Mailed check          29.75        301.90     No  
8           Electronic check         104.80       3046.05    Yes  
9  Bank transfer (automatic)          56.15       3487.95     No  

[10 rows x 21 columns]

- Churn rate is an important factor an organization should focus on as it is directly related to customer satisfaction and revenue genration.
- Churn rate = (No. of customers who have churned * 100) / Total no. customers
- The column 'Churn' here tells us about the customers who have not renewed their contracts.

In [13]:
# Finding count of customers who have churned
df['Churn'].value_counts()

Churn
No     5163
Yes    1869
Name: count, dtype: int64

In [14]:
# Finding churn rate
churn_count = df['Churn'].value_counts()
churn_rate = round(churn_count.iloc[1]*100/churn_count.sum(),2)
print(churn_rate)

26.58


- We can drop 'customerID' as it doesn't seem to serve any purpose in analysis.

In [15]:
# Dropping customerID column
df = df.drop('customerID', axis=1)

### Separating numerical and categorical data to analyze them separately.
- SeniorCitizen, tenure, MonthlyCharges and TotalCharges contain numerical data. Rest contain categorical data.

In [16]:
Numerical_data = ['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges']
Categorical_data = [x for x in df if x not in Numerical_data]

In [17]:
# Exploring categories in categorical data
for i in df:
    if i in Categorical_data:
        print(df[i].value_counts())
        print('-----------------------------')

gender
Male      3549
Female    3483
Name: count, dtype: int64
-----------------------------
Partner
No     3639
Yes    3393
Name: count, dtype: int64
-----------------------------
Dependents
No     4933
Yes    2099
Name: count, dtype: int64
-----------------------------
PhoneService
Yes    6352
No      680
Name: count, dtype: int64
-----------------------------
MultipleLines
No                  3385
Yes                 2967
No phone service     680
Name: count, dtype: int64
-----------------------------
InternetService
Fiber optic    3096
DSL            2416
No             1520
Name: count, dtype: int64
-----------------------------
OnlineSecurity
No                     3497
Yes                    2015
No internet service    1520
Name: count, dtype: int64
-----------------------------
OnlineBackup
No                     3087
Yes                    2425
No internet service    1520
Name: count, dtype: int64
-----------------------------
DeviceProtection
No                     3094
Yes  